ADBDEV-4107 Fix gpmon logging for various types of queries

dnskvlnk opened the pull request:
gpperfmon logs nested queries for `refresh material view`, `create table as` and `create extension` incorrectly, these queries don't have query text and have incorrect submit time. This happened because gpdb didn't fill these field for such types of queries before entering executor. As for the first two, I just added logging functions and now these queries work. As for the `create extension`, I disabled logging for it completely, because of absence of sql strings for separate "create extension" executor commands. We may log only whole sql extension script content for each command that seems to be useless.

Commit:
Enable CTAS queries logging

Commit:
Fix logging for REFRESH MATERIALIZED VIEW

Commit:
Remove extension creation logging

Stolb27 commented on src/backend/commands/extension.c:
```diff
@@ -693,6 +693,16 @@ execute_sql_string(const char *sql, const char *filename)
+	 * such queries weren't logged correctly before this patch, we decided
+	 * to remove this logging entirely
+	 */
+	gp_enable_gpperfmon = false;
```
Can you explain why did you decide to not log nested queries for create extension at all? Why do you talk about log level dependency? Why do we need to care about log level for `create extension` statement?

dnskvlnk replied:
I thought that `warning` log level is enforced even for extension's subqueries, but I see that it's not. I removed this line and reworked extension handling

Stolb27 replied:
We've decided to disable gpperfmon for create extension nested statements as far as we have only extension script content as sql string that is non-informative

Stolb27 commented on src/backend/commands/createas.c:
```diff
@@ -418,6 +420,18 @@ ExecCreateTableAs(CreateTableAsStmt *stmt, const char *queryString,
 
+	if (gp_enable_gpperfmon 
+			&& Gp_role == GP_ROLE_DISPATCH 
+			&& log_min_messages < DEBUG4)
```
Why do we need to care about log level for regular CTAS?

dnskvlnk replied:
Fixed

Commit:
Fix log level for CTAS operations

Commit:
Rework extension creation logging

Commit:
Log only extension's subqueries

Stolb27 commented on src/backend/commands/extension.c:
```diff
@@ -816,6 +818,12 @@ execute_extension_script(Node *stmt,
+
+	if (gp_enable_gpperfmon > DEBUG4)
+	{
+		gp_enable_gpperfmon = false;
```
We've decided to disable gpperfmon for create extension nested statements as far as we have only extension  script content as sql string that is non-informative

Commit:
Disable logging inside extensions

dnskvlnk commented:
I was unable to use `set_config_option` for `gp_enable_gpperfmon` bacause gpdb complains that this guc can be changed only with server restart, so I had to use an old approach with temp variable `orig_gp_enable_gpperfmon`

Stolb27 commented on src/backend/executor/spi.c:
```diff
@@ -2606,91 +2606,78 @@ _SPI_pquery(QueryDesc *queryDesc, bool fire_triggers, int64 tcount)
 		ResetUsage();
 #endif
 
-	bool orig_gp_enable_gpperfmon = gp_enable_gpperfmon;
```
Move this logic to #591

Commit:
Remove junk from extension handling

Commit:
Reset spi.c

Stolb27 commented on src/backend/commands/extension.c:
```diff
@@ -863,6 +864,7 @@ execute_extension_script(Node *stmt,
 		(void) set_config_option("log_min_messages", "warning",
 								 PGC_SUSET, PGC_S_SESSION,
 								 GUC_ACTION_SAVE, true, 0);
+	gp_enable_gpperfmon = false;
```
Should be inside PG_TRY block as I mentioned in the #591

dnskvlnk replied:
Done

Stolb27 reviewed (commented):
> because gpdb forces "warning" logging level for such queries. It means that we can't log just the parent create extension query and ignore extension's contents, we need to either log everything or nothing. Since gpdb didn't log these queries, I decided to disable it completely

1. create extension is utility statement. perfmon ignores them
2. The main reason is absence of sql string for a separate create extension executor commands. We may log only whole sql extesnion script content for each command that seems to be useless.

Commit:
Move gp_enable_gpperfmon assignment inside PG_TRY block

Commit:
Fix gpperfmon logging for various types of queries

gpperfmon logs nested queries for refresh material view, create table
as and create extension incorrectly, these queries don't have query
text and have incorrect submit time. This happened because gpdb didn't
fill these field for such types of queries before entering executor.
As for the first two, I just added logging functions and now these
queries work. As for the create extension, I disabled logging for it
completely, because of absence of sql strings for separate "create extension"
executor commands. We may log only whole sql extension script content for
each command that seems to be useless.

Commit:
Merge branch 'adb-6.x-dev' into ADBDEV-3914-2

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/52876

RekGRpth commented on src/backend/commands/extension.c:
```diff
@@ -908,6 +909,8 @@ execute_extension_script(Node *stmt,
 
 	PG_TRY();
 	{
+		gp_enable_gpperfmon = false;
```
Why don't you want to log the extension creation command by passing the string "create extension ..." to the gpmon_qlog_query_text function?

dnskvlnk replied:
That's an utility command which are not logged by gpperfmon at all, this would lead to inconsistency and require us to change how utility commands are handled because they don't have appropriate data to be passed to gpperfmon

RekGRpth replied:
> they don't have appropriate data to be passed

We can construct string from stmt variable

dnskvlnk replied:
Yes, but it's still inconsistent, moreover, as I remember, gpperfmon support is dropped in 7x and we have our own utility with the same feature

RekGRpth commented on src/backend/commands/matview.c:
```diff
@@ -440,6 +440,17 @@ refresh_matview_datafill(DestReceiver *dest, Query *query,
+		gpmon_qlog_query_text(queryDesc->gpmon_pkt,
+				queryString,
+				application_name,
+				NULL,
```
Why is there no information about resource queues like above?

dnskvlnk replied:
It's about QueryDesc. As I remember, you won't be able to use `gpmon_qlog_packet_init` with QueryDesc created with CreateUtilityQueryDesc

Stolb27 replied:
What? The query desc is created by the regular CreateQueryDesc here.

dnskvlnk replied:
Sorry, misunderstood your question. I added resqueue logging for matview

Commit:
Add resource queue logging for matview

RekGRpth commented on src/backend/commands/matview.c:
```diff
@@ -440,6 +442,17 @@ refresh_matview_datafill(DestReceiver *dest, Query *query,
 
+	if (gp_enable_gpperfmon && Gp_role == GP_ROLE_DISPATCH)
+	{
+		Assert(queryString);
```
Is it redundant?

dnskvlnk replied:
I guess it's better to have this check for a case like in this patch. I'll add this for `create as`

Commit:
Add queryString check for createAs

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/54076

Commit:
Merge branch 'adb-6.x-dev' into ADBDEV-3914-2

Stolb27 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-3914-2

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/54275
